artifact/scripts/executor.cpp

```diff
@@ -0,0 +1,252 @@
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <artifact/scripts/executor.hpp>
+
+#include <algorithm>
+#include <chrono>
+#include <regex>
+#include <string>
+
+#include <common/common.hpp>
+#include <common/expected.hpp>
+#include <common/path.hpp>
+
+
+namespace mender {
+namespace artifact {
+namespace scripts {
+namespace executor {
+
+namespace expected = mender::common::expected;
+
+
+using expected::ExpectedBool;
+
+namespace common = mender::common;
+namespace merror = mender::common::error;
+namespace path = mender::common::path;
+
+
+const int expected_state_script_version {3};
+
+unordered_map<const State, string> state_map {
+	{State::Idle, "Idle"},
+	{State::Sync, "Sync"},
+	{State::Download, "Download"},
+	{State::ArtifactInstall, "ArtifactInstall"},
+	{State::ArtifactReboot, "ArtifactReboot"},
+	{State::ArtifactCommit, "ArtifactCommit"},
+	{State::ArtifactRollback, "ArtifactRollback"},
+	{State::ArtifactRollbackReboot, "ArtifactRollbackReboot"},
+	{State::ArtifactFailure, "ArtifactFailure"},
+};
+
+unordered_map<const Action, string> action_map {
+	{Action::Enter, "Enter"},
+	{Action::Leave, "Leave"},
+	{Action::Error, "Error"},
+};
+
+merror::Error CorrectVersionFile(const string &path) {
+	ifstream vf {path};
+
+	// Missing file is OK
+	if (!vf.is_open()) {
+		return merror::NoError;
+	}
+
+	if (!vf) {
+		auto errnum {errno};
+		return merror::Error(
+			generic_category().default_error_condition(errnum), "Failed to open the version file");
+	}
+
+	string version;
+	vf >> version;
+	if (!vf) {
+		auto errnum {errno};
+		return merror::Error(
+			generic_category().default_error_condition(errnum),
+			"Error reading the version number from the version file");
+	}
+
+	auto exp_version_number {common::StringToInt(version)};
+	if (!exp_version_number) {
+		return executor::error::MakeError(
+			executor::error::VersionFileError,
+			"Error reading the version number from the version file. Is the file in the correct format? error: "
+				+ exp_version_number.error().String());
+	}
+
+	if (exp_version_number.value() != expected_state_script_version) {
+		return executor::error::MakeError(
+			executor::error::VersionFileWrongVersionError,
+			"Unexpected Artifact script version found: " + to_string(exp_version_number.value()));
+	}
+	return merror::NoError;
+}
+
+bool isValidStateScript(const string &file, State state, Action action) {
+	string expression {
+		"(" + state_map.at(state) + ")" + "_(" + action_map.at(action) + ")_[0-9][0-9](_\\S+)?"};
```

Comment:
I don't think you need a regular expression for this and we are trying to avoid them in other places.